website/src/content/docs/tutorials/sca/snr.mdx

```diff
@@ -0,0 +1,540 @@
+
+We thus see that we need to keep track of mean and variances (for a limited
+number of classes). These can be done using a single pass over the data with
+[Welford's online algorithm
+(Wikipedia)](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm).
+
+## A First Try
+
+The SCAAML package implements a version of the online algorithm. It uses pure
+NumPy and is rather conservative with regards to numerical errors (e.g., using
+[Kahan-Babuska sum
+algorithm](https://en.wikipedia.org/wiki/Kahan_summation_algorithm) for
+summing).
+
+When we run the vanilla NumPy algorithm on the `"test"` split of TinyAES dataset
+
+```python
+import numpy as np
+from tqdm import tqdm
+
+from scaaml.stats.snr import SNRSinglePass
+from scaaml.stats.attack_points.aes_128.attack_points import (
+    LeakageModelAES128,
+    SubBytesIn,
+    SubBytesOut,
+)
+from sedpack.io import Dataset
+
+
+# Load the dataset
+dataset = Dataset(dataset_path)
+
+snr = SNRSinglePass(
+    leakage_model=LeakageModelAES128(
+        byte_index=0,
+        attack_point=SubBytesIn(),
+        use_hamming_weight=True,
+    ),
+    db=True,
+)
+
+split = "test"
+
+for example in tqdm(
+        dataset.as_numpy_iterator(
+            split=split,
+            repeat=False,
+            shuffle=0,
+        ),
+        desc=f"[NP] Computing SNR over {split}",
+        total=dataset._dataset_info.splits[split].number_of_examples,
+):
+    snr.update(example)
+
+plt.plot(snr.result)
+plt.savefig("tiny_aes_snr_sbi_0.png")
+```
+
+we get the following SNR (in decibels):
+
+![tiny_aes_snr_sbi_0.png](tiny_aes_snr_sbi_0.png)
+
+## Time Measurements
+
+**This is not a proper benchmark. The tradeoffs for different trace lengths
+might differ.**
+
+On our computer (AMD Ryzen&trade; 9 CPU and SSD) it took around two minutes to
+compute SNR.  We may wonder if this is caused by the data storage being slow or
+by our naive SNR implementation. Let us just iterate through the data and find
+out.
+
+```python
+# Same iteration as before without computation.
+for example in tqdm(
+        dataset.as_numpy_iterator(
+            split=split,
+            repeat=False,
+            shuffle=0,
+        ),
+        desc=f"[NP] Computing SNR over {split}",
+        total=dataset._dataset_info.splits[split].number_of_examples,
+):
+    pass
+```
+
+Just iterating took 8 seconds (around 15 times faster with 3700 iterations per
+second vs previous 250). And spoiler alert we are working on improving our
+implementation to allow even faster iteration.
+
+[Pareto principle (Wikipedia)](https://en.wikipedia.org/wiki/Pareto_principle)
+tells us that we should speed up the computation.
+
+### Possible Improvement Strategies
+
+We have seen that we are bounded by compute. There are many ways how to improve
+this and by no chance can we cover them all here or provide a comprehensive
+benchmarks. Our point is quite opposite sedpack should allow fast iteration of
```

Comment:
typo: `benchmark` (no `s`)

nit (for clarity): `Our point is quite opposite: sedpack should allow ...`